File: workers/AccountingWorker.js
import db from "../db.js";
import { isAlreadyDone, markDone } from "../core/workerIdempotency.js";

const BANK_ACCOUNT_ID = process.env.BANK_ACCOUNT_ID || 3001;

class AccountingWorker {

  static async run(context) {

    const { invoice_id, organization_id } = context;

    // Idempotency — if the full ACCOUNTING step already completed, skip
    if (await isAlreadyDone(invoice_id, organization_id, "ACCOUNTING")) return { nextState: "COMPLETED", reason: "Invoice already processed" };

    // 1. Ensure accrual is posted (has its own internal idempotency check)
    await AccountingWorker.postAccrual(context);

    // 2. Fetch payment schedule
    const res = await db.query(
      `SELECT payment_due_date, payment_status
       FROM invoice_payment_schedule
       WHERE invoice_id = $1
       AND organization_id = $2`,
      [invoice_id, organization_id]
    );

    if (!res.rows.length) {
      throw new Error("Payment schedule missing");
    }

    const { payment_due_date, payment_status } = res.rows[0];

    if (payment_status === "PAID") {
      return { nextState: "COMPLETED", reason: "Invoice already paid" };
    }

    const today = new Date();
    const dueDate = new Date(payment_due_date);

    if (today < dueDate) {
      return { nextState: "ACCOUNTING", reason: "Waiting for payment due date" };
    }

    // 3. Execute payment + clearance journal + markDone — all atomic
    // markDone is inside this transaction so a crash before COMMIT means
    // the record is NOT written → retry will re-run safely
    const client = await db.connect();
    try {
      await client.query("BEGIN");

      await client.query(
        `UPDATE invoice_payment_schedule
         SET payment_status = 'PAID', paid_at = NOW()
         WHERE invoice_id = $1 AND organization_id = $2`,
        [invoice_id, organization_id]
      );

      // Post clearance journal: Dr AP / Cr Bank
      await AccountingWorker.postClearanceJournal(client, context);

      // Populate paid_invoice_registry
      const invoiceRes = await client.query(
        `SELECT data FROM invoice_extracted_data
         WHERE invoice_id = $1 AND organization_id = $2`,
        [invoice_id, organization_id]
      );
      const invoice = invoiceRes.rows[0]?.data || {};

      await client.query(
        `INSERT INTO paid_invoice_registry
           (invoice_id, organization_id, invoice_number, vendor_name, total_amount, paid_at)
         VALUES ($1, $2, $3, $4, $5, NOW())
         ON CONFLICT (organization_id, invoice_number, vendor_name) DO NOTHING`,
        [
          invoice_id,
          organization_id,
          invoice.invoice_number || null,
          invoice.vendor_name || null,
          parseFloat(invoice.total_amount || 0)
        ]
      );

      // markDone inside the same transaction — atomically tied to the work above
      await markDone(invoice_id, organization_id, "ACCOUNTING", client);

      await client.query("COMMIT");
    } catch (err) {
      await client.query("ROLLBACK");
      throw err;
    } finally {
      client.release();
    }

    return { nextState: "COMPLETED", reason: "Payment executed" };
  }

  static async postAccrual(context) {

    const { invoice_id, organization_id } = context;

    // Internal idempotency — accrual has its own check separate from
    // the main ACCOUNTING completion check because it can be posted
    // before the payment clears
    const check = await db.query(
      `SELECT journal_id FROM journal_entries
       WHERE invoice_id = $1 AND organization_id = $2
       AND entry_type = 'ACCRUAL'
       LIMIT 1`,
      [invoice_id, organization_id]
    );

    if (check.rows.length) return;

    const invoiceRes = await db.query(
      `SELECT data FROM invoice_extracted_data
       WHERE invoice_id = $1 AND organization_id = $2`,
      [invoice_id, organization_id]
    );

    const invoice = invoiceRes.rows[0]?.data || {};
    const amount = parseFloat(invoice.total_amount || 0);
    const category = invoice.expense_category || "GENERAL";

    let mappingRes = await db.query(
      `SELECT expense_account_id, ap_account_id
       FROM account_mapping
       WHERE organization_id = $1 AND expense_category = $2`,
      [organization_id, category]
    );

    // GENERAL fallback before throwing
    if (!mappingRes.rows.length) {
      mappingRes = await db.query(
        `SELECT expense_account_id, ap_account_id
         FROM account_mapping
         WHERE organization_id = $1 AND expense_category = 'GENERAL'`,
        [organization_id]
      );

      if (!mappingRes.rows.length) {
        throw new Error(`No account mapping for category: ${category} and no GENERAL fallback`);
      }
    }

    const { expense_account_id, ap_account_id } = mappingRes.rows[0];

    const client = await db.connect();
    try {
      await client.query("BEGIN");

      const entryRes = await client.query(
        `INSERT INTO journal_entries
           (invoice_id, organization_id, entry_type, status, created_at)
         VALUES ($1, $2, 'ACCRUAL', 'POSTED', NOW())
         RETURNING journal_id`,
        [invoice_id, organization_id]
      );

      const journal_id = entryRes.rows[0].journal_id;

      // Dr Expense
      await client.query(
        `INSERT INTO journal_lines (journal_id, account_id, debit_amount, credit_amount)
         VALUES ($1, $2, $3, 0)`,
        [journal_id, expense_account_id, amount]
      );

      // Cr Accounts Payable
      await client.query(
        `INSERT INTO journal_lines (journal_id, account_id, debit_amount, credit_amount)
         VALUES ($1, $2, 0, $3)`,
        [journal_id, ap_account_id, amount]
      );

      await client.query("COMMIT");
    } catch (err) {
      await client.query("ROLLBACK");
      throw err;
    } finally {
      client.release();
    }
  }

  static async postClearanceJournal(client, context) {

    const { invoice_id, organization_id } = context;

    const check = await client.query(
      `SELECT journal_id FROM journal_entries
       WHERE invoice_id = $1 AND organization_id = $2
       AND entry_type = 'CLEARANCE'
       LIMIT 1`,
      [invoice_id, organization_id]
    );

    if (check.rows.length) return;

    const invoiceRes = await client.query(
      `SELECT data FROM invoice_extracted_data
       WHERE invoice_id = $1 AND organization_id = $2`,
      [invoice_id, organization_id]
    );

    const invoice = invoiceRes.rows[0]?.data || {};
    const amount = parseFloat(invoice.total_amount || 0);
    const category = invoice.expense_category || "GENERAL";

    let mappingRes = await client.query(
      `SELECT ap_account_id FROM account_mapping
       WHERE organization_id = $1 AND expense_category = $2`,
      [organization_id, category]
    );

    if (!mappingRes.rows.length) {
      mappingRes = await client.query(
        `SELECT ap_account_id FROM account_mapping
         WHERE organization_id = $1 AND expense_category = 'GENERAL'`,
        [organization_id]
      );
    }

    if (!mappingRes.rows.length) return;

    const { ap_account_id } = mappingRes.rows[0];

    const entryRes = await client.query(
      `INSERT INTO journal_entries
         (invoice_id, organization_id, entry_type, status, created_at)
       VALUES ($1, $2, 'CLEARANCE', 'POSTED', NOW())
       RETURNING journal_id`,
      [invoice_id, organization_id]
    );

    const journal_id = entryRes.rows[0].journal_id;

    // Dr Accounts Payable
    await client.query(
      `INSERT INTO journal_lines (journal_id, account_id, debit_amount, credit_amount)
       VALUES ($1, $2, $3, 0)`,
      [journal_id, ap_account_id, amount]
    );

    // Cr Bank
    await client.query(
      `INSERT INTO journal_lines (journal_id, account_id, debit_amount, credit_amount)
       VALUES ($1, $2, 0, $3)`,
      [journal_id, BANK_ACCOUNT_ID, amount]
    );
  }
}

export default AccountingWorker;

File: python/workers/accounting_worker.py
from db import get_connection
from datetime import datetime
import os

def run(state: dict) -> dict:
    invoice_id = state["invoice_id"]
    organization_id = state["organization_id"]

    # get extracted invoice data
    conn = get_connection()
    cur = conn.cursor()
    cur.execute(
        "SELECT data FROM invoice_extracted_data WHERE invoice_id = %s AND organization_id = %s",
        (invoice_id, organization_id)
    )
    row = cur.fetchone()
    cur.close()
    conn.close()

    if not row:
        return {**state, "next_state": None, "reason": "Extracted data not found"}

    invoice = row[0]
    amount = float(invoice.get("total_amount") or 0)
    category = invoice.get("expense_category") or "GENERAL"

    # check if journal entry already exists — idempotency
    conn = get_connection()
    cur = conn.cursor()
    cur.execute(
        "SELECT journal_id FROM journal_entries WHERE invoice_id = %s AND organization_id = %s LIMIT 1",
        (invoice_id, organization_id)
    )
    existing = cur.fetchone()
    cur.close()
    conn.close()

    if not existing:

        # get account mapping
        conn = get_connection()
        cur = conn.cursor()
        cur.execute(
            """
            SELECT expense_account_id, ap_account_id
            FROM account_mapping
            WHERE organization_id = %s
            AND expense_category = %s
            """,
            (organization_id, category)
        )
        mapping = cur.fetchone()
        cur.close()
        conn.close()

        if not mapping:
            # fallback to GENERAL
            conn = get_connection()
            cur = conn.cursor()
            cur.execute(
                """
                SELECT expense_account_id, ap_account_id
                FROM account_mapping
                WHERE organization_id = %s
                AND expense_category = 'GENERAL'
                """,
                (organization_id,)
            )
            mapping = cur.fetchone()
            cur.close()
            conn.close()

        if not mapping:
            return {**state, "next_state": "EXCEPTION_REVIEW", "reason": "No account mapping found"}

        expense_account_id, ap_account_id = mapping

        # write journal entry + lines atomically
        conn = get_connection()
        cur = conn.cursor()

        try:
            # insert journal entry
            cur.execute(
                """
                INSERT INTO journal_entries
                    (invoice_id, organization_id, entry_type, status, created_at)
                VALUES (%s, %s, 'ACCRUAL', 'POSTED', NOW())
                RETURNING journal_id
                """,
                (invoice_id, organization_id)
            )
            journal_id = cur.fetchone()[0]

            # dr expense account
            cur.execute(
                """
                INSERT INTO journal_lines
                    (journal_id, account_id, debit_amount, credit_amount)
                VALUES (%s, %s, %s, %s)
                """,
                (journal_id, expense_account_id, amount, 0)
            )

            # cr accounts payable
            cur.execute(
                """
                INSERT INTO journal_lines
                    (journal_id, account_id, debit_amount, credit_amount)
                VALUES (%s, %s, %s, %s)
                """,
                (journal_id, ap_account_id, 0, amount)
            )

            conn.commit()

        except Exception as e:
            conn.rollback()
            cur.close()
            conn.close()
            return {**state, "next_state": "EXCEPTION_REVIEW", "reason": f"Journal entry failed: {str(e)}"}

        cur.close()
        conn.close()

    # check payment schedule
    conn = get_connection()
    cur = conn.cursor()
    cur.execute(
        """
        SELECT payment_due_date, payment_status
        FROM invoice_payment_schedule
        WHERE invoice_id = %s AND organization_id = %s
        """,
        (invoice_id, organization_id)
    )
    row = cur.fetchone()
    cur.close()
    conn.close()

    if not row:
        return {**state, "next_state": "EXCEPTION_REVIEW", "reason": "Payment schedule missing"}

    payment_due_date, payment_status = row

    if payment_status == "PAID":
        return {**state, "next_state": "COMPLETED", "reason": "Invoice already paid"}

    today = datetime.now().date()
    due_date = payment_due_date.date() if hasattr(payment_due_date, "date") else payment_due_date

    # not due yet — stay in ACCOUNTING
    if today < due_date:
        return {**state, "next_state": "ACCOUNTING", "reason": f"Waiting for payment due date: {due_date}"}

    # execute payment
    conn = get_connection()
    cur = conn.cursor()
    cur.execute(
        """
        UPDATE invoice_payment_schedule
        SET payment_status = 'PAID', paid_at = NOW()
        WHERE invoice_id = %s AND organization_id = %s
        """,
        (invoice_id, organization_id)
    )

    # write payment clearance journal — Dr AP / Cr Bank
    cur.execute(
        """
        INSERT INTO journal_entries
            (invoice_id, organization_id, entry_type, status, created_at)
        VALUES (%s, %s, 'PAYMENT', 'POSTED', NOW())
        RETURNING journal_id
        """,
        (invoice_id, organization_id)
    )
    journal_id = cur.fetchone()[0]

    cur.execute(
        "INSERT INTO journal_lines (journal_id, account_id, debit_amount, credit_amount) VALUES (%s, %s, %s, %s)",
        (journal_id, 2001, amount, 0)  # Dr AP
    )
    cur.execute(
        "INSERT INTO journal_lines (journal_id, account_id, debit_amount, credit_amount) VALUES (%s, %s, %s, %s)",
        (journal_id, int(os.getenv("BANK_ACCOUNT_ID", 3001)), 0, amount)  # Cr Bank
    )

    conn.commit()
    cur.close()
    conn.close()

    return {**state, "next_state": "COMPLETED", "reason": "Payment executed successfully"}